dev/buildtool/command.py

```diff
@@ -0,0 +1,259 @@
+      raise
+
+  def _do_command(self):
+    """This should be overriden."""
+    # pylint: disable=unused-variable
+    raise NotImplementedError('{0}._do_command is not implemented'.format(
+        self.__class__.name))
+
+
+def _do_call_do_repository(repository, command):
+  """Run the command's _do_repository on the given repository."""
+  # pylint: disable=protected-access
+  logging.info('%s processing %s', command.name, repository.name)
+  try:
+    result = command._do_repository(repository)
+    logging.info('%s finished %s', command.name, repository.name)
+    return result
+  except Exception as ex:
+    maybe_log_exception(
+        '{command} on repo={repo}'.format(
+            command=command.name, repo=repository.name),
+        ex)
+    raise
+
+
+class RepositoryCommandProcessor(CommandProcessor):
+  """And abstract command processor that run a command for each repository."""
+
+  @property
+  def git(self):
+    """Returns GitRunner."""
+    return self.__git
+
+  @property
+  def source_repositories(self):
+    """Returns list of RemoteGitRepository this command should operate on."""
+    if self.__source_repositories is None:
+      self.__source_repositories = self._do_determine_source_repositories()
+    return self.__source_repositories
+
+  @property
+  def source_code_manager(self):
+    """Returns the SourceCodeManager this command uses."""
+    if self.__scm is None:
+      options = self.options
+      self.__scm = SpinnakerSourceCodeManager(
+          self.__git,
+          options.root_path,
+          self.source_repositories,
+          max_threads=self.__max_threads)
+    return self.__scm
+
+  def __init__(self, factory, options, **kwargs):
+    self.__max_threads = kwargs.pop('max_threads', 64)
+    if options.one_at_a_time:
+      logging.debug('Limiting %s to one thread.', factory.name)
+      self.__max_threads = 1
+
+    self.__use_threadpool = kwargs.pop('use_threadpool', False)
+    self.__git = kwargs.pop('git', None) or GitRunner()
+    self.__source_repositories = kwargs.pop('source_repositories', None)
+    self.__scm = None
+    super(RepositoryCommandProcessor, self).__init__(
+        factory, options, **kwargs)
+
+  def filter_repositories(self, source_repositories):
+    """Filter a list of source_repositories using option constriants."""
+    # pylint: disable=unused-argument
+    return source_repositories
+
+  def _do_determine_source_repositories(self):
+    """Determine which repositories this command should operate on."""
+    # pylint: disable=unused-argument
+    raise NotImplementedError(
+        '{0}.determine_source_repositories not implemented'.format(
+            self.__class__.__name__))
+
+  def _do_command(self):
+    self._do_preprocess()
+    result_dict = self.source_code_manager.foreach_source_repository(
+        _do_call_do_repository, self,
+        use_threadpool=self.__use_threadpool)
+    return self._do_postprocess(result_dict)
+
+  def _do_preprocess(self):
```

Comment:
To be extended, right?

Reply:
yea, I clarified instructions in _do_command.
This is a hook so derived classes dont need to implement _do_command rather one of the new _do_ methods depending on where in the processing they want to add behaviors.